File: 프로그래머스/LV2/코딩테스트 연습/2020 카카오 인턴십/수식 최대화/수식 최대화.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

long long solution(string expression)
{
    long long answer = 0;
    string orders[6] = { "*+-", "*-+","+*-", "+-*", "-*+", "-+*" };
    string temp = "";
    vector<long long> nums;
    vector<char> op;

    for (char c : expression)
    {
        if (c == '+' || c == '*' || c == '-')
        {
            op.push_back(c);
            nums.push_back(stoi(temp));
            temp = "";
        }
        else temp += c;
    }
    nums.push_back(stoi(temp));

    for (string order : orders)
    {
        vector<long long> temp_nums = nums;
        vector<char> temp_op = op;
        for (char symbor : order)
        {
            for (int i = 0; i < temp_op.size(); i++) 
            {
                if (symbor == temp_op[i]) 
                {
                    if (symbor == '*') 
                    {
                        temp_nums[i] = temp_nums[i] * temp_nums[i + 1];
                        temp_nums.erase(temp_nums.begin() + i + 1);
                    }
                    else if (symbor == '+') 
                    {
                        temp_nums[i] = temp_nums[i] + temp_nums[i + 1];
                        temp_nums.erase(temp_nums.begin() + i + 1);
                    }
                    else if (symbor == '-') 
                    {
                        temp_nums[i] = temp_nums[i] - temp_nums[i + 1];
                        temp_nums.erase(temp_nums.begin() + i + 1);
                    }
                    temp_op.erase(temp_op.begin() + i--);
                }
            }
        }
        answer = max(answer, abs(temp_nums[0]));
    }
    return answer;
}

File: 프로그래머스/LV2/코딩테스트 연습/2020 카카오 인턴십/수식 최대화/수식 최대화.py
import itertools

def excute_order(order, expr):
    e_last = []
    second = []
    third = expr.split(order[2])
    e = ""
    for t in third:
        second.append(t.split(order[1]))

    for s in second:
        if len(s) == 1:
            e_last.append(str(eval(s[0])))
        else:
            temp = ""
            for c in s:
                c = str(eval(c))
                temp += c + order[1]
            temp = temp.rstrip(order[1])
            e_last.append(str(eval(temp)))

    for f in e_last:
        e += f + order[2]
    e = e.rstrip(order[2])

    return abs(eval(e))

def solution(expression):
    ans = 0
    symbols = ['*', '+', '-']
    orders = itertools.permutations(symbols, 3)
    for order in orders:
        ans = max(ans, excute_order(order, expression))
    return ans